src/runtime/chan.go

```diff
@@ -1,520 +1,329 @@
-	// If this channel operation is not part of a select, then the pointer field of the state holds the data buffer.
-	// If this channel operation is part of a select, then the pointer field of the state holds the receive buffer.
-	// If this channel operation is a receive, then the data field should be set to zero when resuming due to channel closure.
-	t *task.Task
-
-	// s is a pointer to the channel select state corresponding to this operation.
-	// This will be nil if and only if this channel operation is not part of a select statement.
-	// If this is a send operation, then the send buffer can be found in this select state.
-	s *chanSelectState
-
-	// allSelectOps is a slice containing all of the channel operations involved with this select statement.
-	// Before resuming the task, all other channel operations on this select statement should be canceled by removing them from their corresponding lists.
-	allSelectOps []channelBlockedList
+const (
+	chanOperationWaiting = iota // waiting for a send/receive operation to continue
+	chanOperationOk             // successfully sent or received (not closed)
+	chanOperationClosed         // channel was closed, the value has been zeroed
+	chanOperationMask    = 0b11
+)
+
+type chanQueue struct {
+	first *channelOp
 }
 
-// remove takes the current list of blocked channel operations and removes the specified operation.
-// This returns the resulting list, or nil if the resulting list is empty.
-// A nil receiver is treated as an empty list.
-func (b *channelBlockedList) remove(old *channelBlockedList) *channelBlockedList {
-	if b == old {
-		return b.next
-	}
-	c := b
-	for ; c != nil && c.next != old; c = c.next {
-	}
-	if c != nil {
-		c.next = old.next
-	}
-	return b
+// Pus the next channel operation to the queue. All appropriate fields must have
+// been initialized already.
+// This function must be called with interrupts disabled.
+func (q *chanQueue) push(node *channelOp) {
+	node.next = q.first
+	q.first = node
 }
 
-// detach removes all other channel operations that are part of the same select statement.
-// If the input is not part of a select statement, this is a no-op.
-// This must be called before resuming any task blocked on a channel operation in order to ensure that it is not placed on the runqueue twice.
-func (b *channelBlockedList) detach() {
-	if b.allSelectOps == nil {
-		// nothing to do
-		return
-	}
-	for i, v := range b.allSelectOps {
-		// cancel all other channel operations that are part of this select statement
-		switch {
-		case &b.allSelectOps[i] == b:
-			// This entry is the one that was already detached.
-			continue
-		case v.t == nil:
-			// This entry is not used (nil channel).
-			continue
+// Pop the next waiting channel from the queue. Channels that are no longer
+// waiting (for example, when they're part of a select operation) will be
+// skipped.
+// This function must be called with interrupts disabled.
```

Comment:
Is 'interrupts disabled' still accurate when our lord and saviour, multithreading support, arrives? Here and below.

Reply:
Yes, though the comment will need to be updated to say something like "this function must be called with the channel lock held and interrupts disabled".  Interrupts still need to be disabled on microcontrollers. What will happen exactly on a chip like the rp2040 (that has both interrupts and two cores) is something I'll figure out when that happens.